File: android/app/src/main/java/com/mathgate/GateStatus.kt
package com.mathgate

import java.time.ZoneId

/** Result of asking "has a Math Academy task been completed in the current period?". Fail closed: only Done unlocks. */
sealed class GateStatus {
    abstract val allowsAccess: Boolean

    data class Done(val doneUntilMillis: Long, val tasksCompleted: Int, val xp: Int) : GateStatus() {
        override val allowsAccess = true
    }

    data class NotDone(val periodStartMillis: Long, val nextResetMillis: Long, val checkedAtMillis: Long) : GateStatus() {
        override val allowsAccess = false
    }

    /** Network unreachable (no connectivity, DNS failure, timeout). */
    data class Offline(val message: String) : GateStatus() {
        override val allowsAccess = false
    }

    /** No credentials saved, wrong credentials, login backoff active, or Keystore decrypt failure. */
    data class AuthFailed(val reason: String) : GateStatus() {
        override val allowsAccess = false
    }

    /** Math Academy answered but not in the shape we expect (API change, HTML instead of JSON, ...). */
    data class UnexpectedResponse(val detail: String) : GateStatus() {
        override val allowsAccess = false
    }

    /** Nothing fetched yet. */
    object Unknown : GateStatus() {
        override val allowsAccess = false
        override fun toString() = "Unknown"
    }
}

data class GateRequest(
    val periodStartMillis: Long,
    val nextResetMillis: Long,
    val zone: ZoneId,
    val forced: Boolean,
)

/** Blocking call; only ever invoked on the coordinator's IO executor. */
interface GateApi {
    fun fetch(request: GateRequest): GateStatus
}

data class Credentials(val username: String, val password: String)

File: ios/MathGateKit/Sources/MathGateKit/GateStatus.swift
import Foundation

/// Result of asking "has a Math Academy task been completed in the current period?".
/// Fail closed: only `.done` unlocks.
public enum GateStatus: Equatable, Sendable {
    case done(doneUntil: Date, tasksCompleted: Int, xp: Int)
    case notDone(periodStart: Date, nextReset: Date, checkedAt: Date)
    /// Network unreachable (no connectivity, DNS failure, timeout).
    case offline(String)
    /// No credentials saved, wrong credentials, login backoff active, or Keychain read failure.
    case authFailed(String)
    /// Math Academy answered but not in the shape we expect (API change, HTML instead of JSON).
    case unexpectedResponse(String)
    /// Nothing fetched yet.
    case unknown

    public var allowsAccess: Bool {
        if case .done = self { return true }
        return false
    }
}

public struct Credentials: Equatable, Sendable {
    public let username: String
    public let password: String

    public init(username: String, password: String) {
        self.username = username
        self.password = password
    }
}

public struct GateRequest: Equatable, Sendable {
    public let periodStart: Date
    public let nextReset: Date
    public let now: Date

    public init(periodStart: Date, nextReset: Date, now: Date) {
        self.periodStart = periodStart
        self.nextReset = nextReset
        self.now = now
    }
}

public protocol GateAPI: Sendable {
    func fetch(_ request: GateRequest) async -> GateStatus
}
